src/util/jsonLog.js

```diff
@@ -0,0 +1,89 @@
+// @flow
+
+import stringify from "json-stable-stringify";
+import fs from "fs-extra";
+import * as C from "./combo";
+
+/**
+ * JsonLog tracks and serializes append-only logs of JSON objects.
+ *
+ * At its heart, it's basically a simple wrapper around an array, which
+ * enforces the rule that items may be appended to it, but never removed.
+ *
+ * It also provides serialization logic. While the log is an array of JSON
+ * objects and thus a JSON object itself, we don't want to store it on disk as
+ * a stringified JSON object. That would lose the structure of an append-only
+ * array, as every change would result in a rewrite of a single massive line of
+ * text, and would be inefficient in Git storage, and make for ugly diffs and
+ * merge conflicts.
+ *
+ * Instead, we store each item on its own line in the output format, using
+ * `stringify` on the individual JSON objects. We support // comment syntax,
+ * and the toString and writeJsonLog methods allow you to provide prefix
+ * comments; we recommend putting prefix comments like
+ * ["AUTOGENERATED FILE", "do not manually edit"].
+ *
+ * The JsonLog is always parsed using a Combo.Parser, which ensures type safety
+ * at runtime.
+ */
+export class JsonLog<T: C.JsonObject> {
+  _items: T[];
+
+  constructor() {
+    this._items = [];
+  }
+
+  append(items: $ReadOnlyArray<T>): JsonLog<T> {
+    this._items = [...this._items, ...items];
+    return this;
+  }
+
+  items(): $ReadOnlyArray<T> {
+    // Make a copy, guarding against unsafe mutation, and avoiding potential
+    // confusion when the client adds more items to the log after requesting
+    // the items.
+    return this._items.slice();
+  }
+
+  toString(prefixComments: ?$ReadOnlyArray<string>): string {
+    const stringified = this._items.map((x) => stringify(x)).join("\n");
```

Comment:
This will surely cause memory problems when the log has a substantial size. Which I think would defeat the purpose of having ndjson. If we're not considering a memory bottleneck, then why not write the entire log as a single JSON array?

Reply:
I'm not worried about in-memory size, since the size of RAM on a users machine is much, much larger than the max file size limitation on GitHub (~gigabytes vs ~100MB), we will need to switch to a very different solution before memory becomes an issue. However, if we write the log as a JSON array, we'll get bad repository bloat on the Git side, because appending a single event would force the ledger to be an entirely new blob. However, if we use ndjson, then [Git packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) will deduplicate most of the log for us, across commits.

Reply:
I think it should be a concern moving forward though.

For dogfooding this may not be our first bottleneck, but the "git hosting" constraints can be avoided switching to Git LFS, IPFS or just keeping on regular filesystems. Today we're not able to run SC on larger projects as it will crash due to memory constraints. 16GB ram processes are the upper limit (regardless of actual ram + swap available) and I've hit it for tests with larger projects.

Streaming ndjson would be my recommendation, but the point remains. We can get the git optimizations with plain JSON. If we're writing our own serializer, might as well wrap it in `[\n` + `\n]\n` and write `,\n` at the end of each row to have plain old JSON on easy to compress rows.

Reply:
Regarding memory--the part of the system that hits the memory bottleneck (serializing huge graphs and cred results) has size `O(num weeks * num nodes)`. This log is being used for data that has size `O(num weeks * num users)`, which is at least 2 orders of magnitude smaller. So I'm not worried about memory pressure on this module right now; when/if we do find a bottleneck, we can improve the implementation without needing to change the output format.

(Also--this ledger data is wholly owned by SourceCred, if we need to upgrade it in the future we can still do that.)

I do like the idea of just writing plain JSON rather than ndjson. It's a neat trick where do still get the clean diffs, and we don't have the complexity of doing something that isn't regular JSON. So I changed it to your implementation.

Reply:
> (Also--this ledger data is wholly owned by SourceCred, if we need to upgrade it in the future we can still do that.)

I wrote a small essay on that before, but in my experience this isn't how upgrading works with event logs. https://github.com/sourcecred/sourcecred/pull/1697#discussion_r409020751
> For better or worse you will be supporting these event models until the end of time. I disagree with @decentralion here, with regards to "private data" and upgrade paths. These events are your source of truth and updating them after the fact is a breach of contract. Even re-encoding the same data with the same meaning into a new format should be frowned upon. If this were on-chain data, this would rightfully be made infeasible to do.

However switching from JSON arrays to NDJSON is something I think can be done, because they're both encodings of `Array<T>` that are easy to prove are equivalent. Plus backward compatibility would not be a big deal.

Reply:
For posterity/context, this change was reversed in #2075.